Setup api dll

vbfox opened the pull request:
Add a few functions from SetupApi and their helpers.

I also added a `PinnedStruct<T>` class in core to handle cases where the api essentially expect a `Foo?` where `Foo` is a structure.

The class take care of either providing `IntrPtr.Zero` if null is passed or boxing the value and pinning it otherwise.

I never needed it but that might be one of the class where adding a destructor could be justified, what do you think ?

vbfox committed:
Run AddNewLibrary.ps1 for SetupApi.dll

vbfox committed:
Add SetupApi to solution

vbfox committed:
Change R# rule for spaces after casts

vbfox committed:
Add DllImport for SetupApi

vbfox committed:
Add PinnedStruct to Windows.Core

vbfox committed:
Synchronize private static readonly between R# and StyleCop

vbfox committed:
Add helpers for SetupApi

vbfox committed:
Fix a StyleCop issue

vbfox committed:
Fix ambigous cref in xml doc

vbfox committed:
Add nuget reference to Kernel32 for SetupApi

vbfox committed:
Remove Portable and Shared projects for SetupApi

The API are marked as Desktop only

vbfox commented:
Also I have been using HResult/ComException to check for errors but as the lib is desktop only Win32Exception is available.

(But maybe that could be changed latter with a single unified way to handle LastError in the whole lib)

AArnott commented on src/SetupApi.Desktop/SetupApi.cs:
```diff
@@ -0,0 +1,195 @@
+        [DllImport(nameof(SetupApi), SetLastError = true)]
+        public static extern SafeDeviceInfoSetHandle SetupDiGetClassDevs(
+            IntPtr classGuid,
+            [MarshalAs(UnmanagedType.LPStr)] string enumerator,
```
I understand that `LPTStr` is the default marshaling for string. I'm guessing `LPStr` means always use ANSI instead of Unicode. Would this attribute then be equivalent to simply adding `CharSet = CharSet.Ansi` to the `DllImport` attribute? I'm thinking declaring it at the method level rather than the parameter level would be easier to read.

vbfox replied:
It's actually just me trusting code I had from another project it's actually a `LPTStr` so the default value and doesn't need specifying

AArnott commented on src/SetupApi.Desktop/SetupApi.cs:
```diff
@@ -0,0 +1,195 @@
+        /// error, <see langword="false" /> is returned and the error code for the failure can be retrieved by calling
+        /// <see cref="Marshal.GetLastWin32Error" />.
+        /// </returns>
+        [DllImport(nameof(SetupApi), SetLastError = true, CharSet = CharSet.Auto)]
```
Do you happen to be able to explain to me what `CharSet.Auto` means? .NET is fundamentally unicode, so when would it ever use Ansi? Or is it auto in the sense that it somehow magically figures out whether the native method expects ansi or not? Is this where .NET will add an A or W suffix to the method? (and if so, when would it ever add A)

vbfox replied:
i'll remove auto there as anyway no strings are used and there doesn't seem to be any A or W version.

But the problem is the default charset value in .Net.

This value drive both what is called (A/W) and how string are marshaled if they don't have a more specific attribute.

The problem here is that the default value is Ansi, not auto, so any method declared without it that contain strings will have it's ansi version called. ( https://msdn.microsoft.com/en-us/library/7b93s42f(v=vs.110).aspx should display a big red warning about that)

It's a bug I see often, especially in code comming from the US.
_A_ versions are used (as it's the default) by .Net app and it work as long as you don't interact with data that aren't in the current ansi range. But for users of character sets that don't easily map ton ansi (china, japan) the program give gliberish results.

It's not important for `SafeDeviceInfoSetHandle` (The method you discussed just before) as the enumerator is actually in a very specific syntax, but actually this function would benefit from having it's CharSet specified.

But it's true that "Auto" or "Ansi" would not be so good choice in this library, first because Auto is acually not supported in the PCL and also as the minimum framework of 4.0 mean windows XP so W versions would always be used anyway.

AArnott commented on src/Windows.Core/PinnedStructExtensions.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace PInvoke
+{
+    public static class PinnedStructExtensions
```
XML doc comments on the type and all its members would be useful. Can you add?

vbfox replied:
Yep

AArnott commented on src/Windows.Core/PinnedStruct{T}.cs:
```diff
@@ -0,0 +1,89 @@
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException("PinnedStruct");
```
use `nameof` here?

vbfox replied:
Good catch

AArnott commented on src/Windows.Core/PinnedStruct{T}.cs:
```diff
@@ -0,0 +1,89 @@
+                this.gcHandle = GCHandle.Alloc(box, GCHandleType.Pinned);
+                this.intPtr = this.gcHandle.Value.AddrOfPinnedObject();
+            }
+            else
```
I would lose the else block here. In .NET, a class's fields are initialized to their default values automatically.

vbfox replied:
Ok it was mostly here for clarity :)

AArnott commented on src/Windows.Core/PinnedStruct{T}.cs:
```diff
@@ -0,0 +1,89 @@
+    /// Allocate a structure on the heap and pin it in place, allowing a controlled release (Via Dispose).
+    /// </summary>
+    /// <typeparam name="T">Type of the structure that will be pinned.</typeparam>
+    public class PinnedStruct<T> : IDisposable
```
This is a great idea. I've used IntPtr in method signatures for lack of this already so I can fix up my code after this merges.
I wonder if `StrongBox<T>` might just happen to work for us. Have you tried it?

vbfox replied:
I didn't know about it.

It would make the code slightly cleaner as it avoid the cast in `Value` but i'll need to experiment on how after pinning the `StrongBox<T>` instance I can get a pointer to the inner `Value` field. Maybe for another PR :)

AArnott commented on src/Windows.Core/PinnedStruct{T}.cs:
```diff
@@ -0,0 +1,89 @@
+        {
+            if (this.gcHandle.HasValue && !this.disposed)
+            {
+                this.gcHandle.Value.Free();
```
Yes, it looks like this is a good case to have a Finalizer. GCHandle can't do it since it's just a struct. So we should help it out.

AArnott commented:
Done reviewing. Overall it looks good. :)

vbfox committed:
Fix a few comments from the review (#9)

vbfox committed:
Add a few more <see>

vbfox committed:
Add Charset to SetupDiGetDeviceInterfaceDetail

vbfox commented:
So from the changes :
* I added CharSet.Unicode to SetupDiGetClassDevs as SetupDiGetClassDevsA exists and there is no reason to use it (Same for SetupDiGetDeviceInterfaceDetail)
* Removed it from SetupDiEnumDeviceInfo as it doesn't have a A/W variant
* I added comments to PinnedStruct & friends
* Didn't use `StrongBox<T>` as  I can't see an easy way to get the field offset, i'll investigate when I have some free time :)

vbfox commented:
Also added the finalizer.

And I think that's all, did I forget a comment ?

AArnott commented:
Looks good. I'd like to see the full disposable pattern in PinnedStruct but I can fix that up after merging.
